### ISSUE TITLE
nbdserver with tlog : only Flush the operation after it get Flushed by tlog

### ISSUE THREAD

**Issue by iwanbk** [g8os/blockstor#153]
part of #126

**Comment by iwanbk** [comment]
How is our progress on this issue?

**Comment by GlenDC** [comment]
Slow but steady. Was doing it in between other things. As it also involved a lot about the tlog logic, I was also going through the entire flow and different paths that happen in the `nbd.Backend <-> tlogclient <-> tlogserver` logic, etc.

In order to resolve this issue as well as #178, as the solution for this will also resolve the other, I was thinking along the following lines:

+ the tlogclient should probably also keep track of the `lastReceivedFlushSequence`, and also somehow `theLastSentSequence`, so it can compare these values to know if all blocks are flushed or not;
+ I think it would be great if the client had a flush function (to force the flushing and sending of blocks more or less), it should do the following:
   1) force send all buffered blocks on the tlogclient side to the server (at least try), if this fails, the flushing fails for sure;
   2) for the next and last step of this function I see 2 options:
       a) either we now wait (_with_ a timeout) until the last expected flush sequences matches our expectations, and thus we know that all sent blocks have been flushed;
       b) on top of going for a, we could perhaps also allow the client somehow to give the server a push, and to request a flush right now, rather than wait until the flushTime or flushSize has been reached, this might be able to speed things up. However it would mean that instead of just accepting block messages, we would need to accept a dynamic typed message, which could be a block, but could be also something differently like a flush request, and whatever else we come up with in the future;

All this does mean, that the nbdserver flushing of the actual block data will only ever be attempted, only if the tlogserver has been flushed. If that behavior is desired (I was not in the original design meetings, so I don't know this), than that is fine I guess. We could of course also wait for the tlog to flush (using the steps described above), but even flush the nbdserver in case the tlog flushing failed. That would mean that at least the nbdserver vdisk is usable for now. No tlogreplay would be available though, so perhaps that's not desired and thus we do want to fail it. Just putting this out here, as I don't know what the original requirements around this were.

Of course if the `nbd.Backend` has no tlogclient (which is possible if no tlogrpc address is given), which is the case for at least 1 vdisk type AFAIK, we can just flush the storage directly and none of this matters.

**Comment by GlenDC** [comment]
Today however I won't have time anymore to do an actual implementation. But would be good to talk to you @iwanbk about this, and combine our knowledge of the 2 modules (nbdserver and tlog), to get to a decent solution. Doesn't have to be perfect yet, but should have at least decent foundations.

**Comment by iwanbk** [comment]
> the tlogclient should probably also keep track of the lastReceivedFlushSequence

It is very easy to do in tlogclient, but i don't think that it is a good idea.
Because it means that nbdserver need to check this value regularly. (polling is usually not a good idea).

I suggest nbdserver to check the response it received from `tlogclient.Recv`, if the status is `BlockStatusFlushOK`, the sequences are all flushed block sequences (see the example in send_tlog example). By knowing the flushed blocks, nbdserver can flushed the corresponding operation.

> and also somehow theLastSentSequence

While it might be valid for now, i don't think lastSeqSequence can be used when we use more than one connections.

>  on top of going for a, we could perhaps also allow the client somehow to give the server a push, and to request a flush right now

yes, someone (i think michal) already suggested this in (now inactive) tlog telegram group, and everyone seems agree.
I think we can only add this feature in next milestone.
For now, we can adjust 'FlushTime' to small value to meet nbd expectation. 
Setting `FlushTime` to X  is almost similar as nbdserver wait for X seconds before forcing the tlog to flush all blocks it has.

>  but even flush the nbdserver in case the tlog flushing failed. That would mean that at least the nbdserver vdisk is usable for now. No tlogreplay would be available though, so perhaps that's not desired and thus we do want to fail it

Once we know tlogreplay won't give valid result, we might not want to use tlogserver anymore for that vdisk.
But of course that is depend on our requirement.
But for now, i suggest that we make the vdisk not usable at all if the tlog failed. So, we and the user can be 100% aware of it.

**Comment by iwanbk** [comment]
> For now, i suggest that we make the vdisk not usable at all if the tlog failed. So, we and the user can be 100% aware of it and fix what code/design that need to be fixed.

And of course the best solution is for nbdserver to only flush after it getting flushed by tlog.
But not sure how it will affect the local disk managed by nbdclient.

**Comment by GlenDC** [comment]
> And of course the best solution is for nbdserver to only flush after it getting flushed by tlog.
But not sure how it will affect the local disk managed by nbdclient.

To be fair, for nondeduped storage it [flushing] doesn't mean anything, as nondeduped storage writes directly to ardb, and the flush function for that type of storage doesn't do anything. For deduped storage, the actual data (blocks) are also written directly. The metadata is only written if a) the cache is full, or b) if we flush. So really, most of the data is already written anyhow (aka flushed), so I think it makes sense to just always make that nbd flush complete (as it will already be like mostly flushed anyhow), even if the tlog one fails (due to a bad connection or whatever).

But even though we flush, we could still return an error, so at least the nbd client would disconnect, and know something is wrong.

**Comment by GlenDC** [comment]
> I suggest nbdserver to check the response it received from tlogclient.Recv, if the status is BlockStatusFlushOK, the sequences are all flushed block sequences (see the example in send_tlog example). By knowing the flushed blocks, nbdserver can flush the corresponding operation.

Wouldn't work, as I said, the flushing is just some partial metadata left over that was still in cache. Without big changes to the nbdserver design this will not change.

But the problem (the one that I wanted to address in issue #178) is that we need to keep the client (that is the nbdserver) open, or else the tlogserver stops handling that connection. As it wants to send responses (at least in the current design).

So when the Backend Flush method is called, this is the _last_ moment the tlogclient has to make sure all blocks are sent to the server. In theory we could exit after that, but we could (and perhaps should) also wait so all blocks are actually flushed in the tlogserver.

**Comment by GlenDC** [comment]
> I suggest nbdserver to check the response it received from tlogclient.Recv, if the status is BlockStatusFlushOK, the sequences are all flushed block sequences (see the example in send_tlog example). By knowing the flushed blocks, nbdserver can flush the corresponding operation.

I was already planning to get the `lastReceivedFlushSequence` from the received `BlockStatusFlushOK` responses, I did see that. Like I said, I went carefully through the tlog code, to get as far as I did.

But yeah this plans does fail if you bring multiple connections. In which case the only thing I can think about, is again the force flush message, which could be failed by the server for any reason (including the reason that it doesn't have all blocks yet (it knows because it has some blocks waiting in its tree collection)).

But you already seem to have pushed that to a alpha-4 issue (#188). So dunno.

Either way, it is clear to me that the code as it is today, won't allow us to have the tlogclient wait for the tlogserver to be done flushing every block sent to it.

**Comment by iwanbk** [comment]
> For deduped storage, the actual data (blocks) are also written directly

OK, no problem with it.

>  the flushing is just some partial metadata left over that was still in cache

Then, i think we need to do this this only after tlog flush, no?

> So when the Backend Flush method is called, this is the last moment the tlogclient has to make sure all blocks are sent to the server. In theory we could exit after that, but we could (and perhaps should) also wait so all blocks are actually flushed in the tlogserver.

Yes, this is what we need to wait for tlog flush.
So, i think it is not `could` or `should` but `must`.

What make sure we need to do above are:
- tlog replay need to call backend.Flush
- Write-ahead logging concept.

**Comment by iwanbk** [comment]
> Either way, it is clear to me that the code as it is today, won't allow us to have the tlogclient wait for the tlogserver to be done flushing every block sent to it.

No, we can, although not as fast as you might wanted.

If we set the `FlushTime` to 3 seconds, tlog will Flush all **ordered blocks** it has, 3 seconds after last flush.
It the same as this pseudocode in nbd:
- wait for 3 seconds to receive `BlockStatusFlushOK` for sequences we want to flush
- if we don't get it, send "force flush"

**Comment by GlenDC** [comment]
> Then, i think we need to do this this only after tlog flush, no?

Could be. But I hope you do realise that some data (blocks) which have already been written to external storage (aka flushed), have not yet been flushed by the tlog. So the amount of data that is left to be flushed by the nbdserver at the point of the `backend.Flush` call, might be less then the amount of blocks that the tlog still has to flush. As the tlogserver might be really behind on schedule, perhaps even due to a very bad failure case scenario.

> tlog replay need to call backend.Flush

Tlog replay already calls backend.Flush.

> Write-ahead logging concept.

what do you mean with this?

**Comment by GlenDC** [comment]
> If we set the FlushTime to 3 seconds, tlog will Flush all ordered blocks it has, 3 seconds after last flush.

This flush time is set via a command line argument when starting the server, there is no way at the moment for the client to know about this value.

> if we don't get it, send "force flush"

That would be issue #188?

**Comment by iwanbk** [comment]
```
tlog replay need to call backend.Flush
Tlog replay already calls backend.Flush.

Write-ahead logging concept.
what do you mean with this?
```
sorry if my statement was not clear.
what i meant, these existing things make me sure that nbdserver can Flush only after tlog Flush it:
- backend.Flush -> well, after i think about it again, it seems that it only weak correlation
- write-ahead logging : https://en.wikipedia.org/wiki/Write-ahead_logging "In a system using WAL, all modifications are written to a log before they are applied"

```
If we set the FlushTime to 3 seconds, tlog will Flush all ordered blocks it has, 3 seconds after last flush.
This flush time is set via a command line argument when starting the server, there is no way at the moment for the client to know about this value.

if we don't get it, send "force flush"
That would be issue #188?
```
sorry again if my statement was not clear (again :D).
So, because you seems to really need "force flush", you must already have a plan to use it, right?
If the plan is according to my pseudocode above, then (for now) setting FlushTime to short value will give similar effect to that code.

But if my suggestion not working, and you really need "force flush" right now, i'll do my best to implement it in this milestone.

Let me know if all above suggestion doesn't make sense, i'll dig nbdserver code deeper to get better view on this matter.

Basically,all of my suggetions/proposals  are in order to achieve this goal: when we use tlogserver, then we need to make sure that tlogserver provide strong consistency.
Otherwise, we can simply use ardb slave which is simpler and faster.

**Comment by iwanbk** [comment]
So, i have quick checked the lba, backend, and deduped code, and send a PR (#190, feel free to quickly reject it if you think it doesn't make sense).

To achieve (my version of ) strong consistency, nbdserver need to do one of these things for all `WriteAt` and `WriteZeroesAt`  operation(lets call  a `WriteAt` / `WriteZeroesAt`  operation a 'transaction'):
- only store that transaction in memory, identified by `tlogCounter` (in nbdserver term) or `sequence` (in tlog term).  Only really execute that transaction after that transaction successfully flushed by the tlog

- it is the same like current code, but still need to store that transaction in memory. If tlogserver failed to flush that transaction, then `undo` that transaction. But i don't have any idea at all to do this `undo` :D 

I prefer the 1st solution, because undo seems tricky.
If you think that tlog flush could take a long time, we have these:
- set FlushTime as short as possible
- add "force flush" -> need to be implemented.

**Comment by GlenDC** [comment]
> only store that transaction in memory, identified by tlogCounter (in nbdserver term) or sequence (in tlog term). Only really execute that transaction after that transaction successfully flushed by the tlog

Not sure if storing it in memory is at all desirable. As this could take a lot of Gigs of memory if we're writing a big image, before we ever flush. Fairly sure we don't have the RAM for that available.

If we want to store it somewhere else first, based on some sequence, we might want to store it in the ardb already, and more it over from there.

No matter if we store it in RAM or ARDB, things would get quite tricky (if not ill-designed), as we often require quite shortly after that to `Get` that freshly written data, as part of a `Merge`.

So not sure if I can see a solution in these suggestions.

**Comment by iwanbk** [comment]
> As this could take a lot of Gigs of memory if we're writing a big image, before we ever flush. Fairly sure we don't have the RAM for that available.

One block is 4KB, default `FlushSize` value of tlog is 25. Which means tlog is going to try to flush it every 100KB. It is true that we will need more than 100KB of buffer, but Gigs is exaggerated.


> No matter if we store it in RAM or ARDB, things would get quite tricky (if not ill-designed), as we often require quite shortly after that to Get that freshly written data, as part of a Merge.

You're right

**Comment by GlenDC** [comment]
OK, after thinking about it (a lot) more, and playing around with some ideas and thoughts, I have the following proposal:

+ there will be a new type of storage, called `tlogStorage`, which wraps around a `deduped`/`nonDeduped` storage;
+ the `ardb.Backend` will no longer do any tlog stuff;
+ the `TlogClient` will instead be used from within the `tlogStorage`;
+ the `tlogStorage` will first keep modified content in a sequence-cache, and only write the modified content once it is stored (flushed) properly in the tlogServer;

Here is a flow graph that summarizes the `tlogStorage` as I have it in mind: 

![tlog_storage](https://cloud.githubusercontent.com/assets/3900482/26517062/ad36c6e2-4256-11e7-882f-a6e26b5829b9.png)

This would give the strong consistency you are looking for @iwanbk.

If this sounds alright to you, I'll implement it all this weekend, which should resolve this issue as well as issues #178 and #126. And with that also complete milestone alpha-3.

**Comment by iwanbk** [comment]
> the tlogStorage will first keep modified content in a sequence-cache, and only write the modified content once it is stored (flushed) properly in the tlogServer;

In memory or some external storage?

> 
> This would give the strong consistency you are looking for @iwanbk.
> 
> If this sounds alright to you, I'll implement it all this weekend,

Can't check your graph today, but it sounds good.

**Comment by GlenDC** [comment]
> In memory or some external storage?

For now it's in-memory, but nothing would stop us from storing it in some external storage (ardb I guess). Storing it in external storage would make it slower though, and I don't think there is a need for it. Either way, the design of my idea doesn't depend on the distinction.

**Comment by GlenDC** [comment]
A current WIP version can be found at: https://github.com/g8os/blockstor/pull/209

It's pretty close to be finished I think. There is still 1 big issue related to the flushing (which doesn't seem to be really properly happening, waiting on the tlogserver to be flushed is pretty flawed, or I'm missing something, either way it will be good to implement more control over the server's behavior such as force flushing and have a more predictable way of waiting because of it). But as you noted before, it should be working with simply waiting, so there might be some other funny business going on.

I don't think any of the fixes required to finish this implementation will really change the idea of this implementation though.

I also posted an up-to-date version of the tlogStorage's flowgraph in that PR (#209).

**Comment by GlenDC** [comment]
Will continue working on this on Monday. For now I need rest. My brain is exhausted.

The milestone is a bit delayed by this issue, but that only shows the importance of involving us more in the roadmap, the global picture, and the plans in general. Helps a lot in the design behind the codebases.
